util/merkle-mountain-range/src/mmr.rs

```diff
@@ -0,0 +1,195 @@
+            .collect::<Result<Option<_>>>()?
+            .ok_or(Error::InconsistentStore)?;
+        while rhs_peak_elems.len() > 1 {
+            let right_peak = rhs_peak_elems.pop().expect("pop");
+            let left_peak = rhs_peak_elems.pop().expect("pop");
+            rhs_peak_elems.push(M::merge(&right_peak, &left_peak));
+        }
+        Ok(rhs_peak_elems.pop())
+    }
+
+    pub fn gen_proof(&self, mut pos: u64) -> Result<MerkleProof<T, M>> {
+        let mut proof: Vec<T> = Vec::new();
+        let mut height = 0;
+        while pos < self.mmr_size {
+            let pos_height = pos_height_in_tree(pos);
+            let next_height = pos_height_in_tree(pos + 1);
+            let (sib_pos, next_pos) = if next_height > pos_height {
+                // implies pos is right sibling
+                let sib_pos = pos - sibling_offset(height);
+                (sib_pos, pos + 1)
+            } else {
+                // pos is left sibling
+                let sib_pos = pos + sibling_offset(height);
+                (sib_pos, pos + parent_offset(height))
+            };
+            if sib_pos > self.mmr_size - 1 {
+                break;
+            }
+            proof.push(
+                self.batch
+                    .get_elem(sib_pos)?
+                    .ok_or(Error::InconsistentStore)?,
+            );
+            pos = next_pos;
+            height += 1;
+        }
+        // now we get peak merkle proof
+        let peak_pos = pos;
+        // calculate bagging proof
+        let peaks = get_peaks(self.mmr_size);
+        if let Some(rhs_peak_hash) = self.bag_rhs_peaks(peak_pos, &peaks[..])? {
+            proof.push(rhs_peak_hash);
+        }
+        let lhs_peaks: Vec<_> = peaks
+            .iter()
+            .filter(|&&p| p < peak_pos)
+            .map(|&p| self.batch.get_elem(p))
+            .rev()
+            .collect::<Result<Option<_>>>()?
+            .ok_or(Error::InconsistentStore)?;
+        proof.extend(lhs_peaks);
+        Ok(MerkleProof::new(self.mmr_size, proof))
+    }
+
+    pub fn commit(self) -> Result<()> {
+        self.batch.commit()
+    }
+}
+
+#[derive(Debug)]
+pub struct MerkleProof<T, M> {
+    mmr_size: u64,
+    proof: Vec<T>,
+    merge: PhantomData<M>,
+}
+
+impl<T: PartialEq + Debug, M: Merge<Item = T>> MerkleProof<T, M> {
+    pub fn new(mmr_size: u64, proof: Vec<T>) -> Self {
+        MerkleProof {
+            mmr_size,
+            proof,
+            merge: PhantomData,
+        }
+    }
+
+    pub fn verify(&self, root: T, mut pos: u64, elem: T) -> Result<bool> {
+        let peaks = get_peaks(self.mmr_size);
+        let mut sum_elem = elem;
+        let mut height = 0;
+        for proof in &self.proof {
+            if peaks.contains(&pos) {
+                sum_elem = if Some(&pos) == peaks.last() {
+                    M::merge(&sum_elem, &proof)
+                } else {
+                    pos = *peaks.last().expect("must exists at least one peak");
```

Comment:
bug? should it be peaks.first()?

Reply:
It's a trick, the elem maybe belongs to a left peak, so after it's pack with right peaks, the remain proofs are always left peaks. set the pos as last peak to make the if condition always true.

I' ll add some comments.